### ISSUE TITLE
Enable health checks to report holistically

### ISSUE THREAD

**Issue by elnyry** [mojaloop/project#796]
## **Goal**:
As a Hub Operator
I want to monitor the health of mojaloop Switch services in more detail
so that rather than a simple up/down info on a service, a more accurate reporting of health is available, that includes metadata such as version numbers, DB connectivity.

**Tasks**:
- [x] Define a standard model for health
- [x] Health services end-points to be refactored to report
  - [x] Overall status
  - [x] The version number of the service (from package.json)
  - [x] Backend connectivity (as relevant)
    - [x] DB
    - [x] Kafka
~- [ ] Sidecar~ Will be refactored soon, so not worth the effort atm
- [ ] Apply the above for
  - [x] ml-api-adapter
  - [x] central-ledger
  - [ ] central-settlement
  - [x] central-event-processor
  - [x] email-notifier

**Acceptance Criteria**:
- [x] Health endpoint model submitted to the DA, is reviewed by the team
- [x] Health end-points on services report health more holistically
- [x] This is applied to ml-api-adapter to start with and then to the rest of the services

## **Pull Requests**:
- [x] https://github.com/mojaloop/documentation/pull/67
~- [ ] https://github.com/mojaloop/central-ledger/pull/285~ Needs to be revisited
- [x] https://github.com/mojaloop/central-services-shared/pull/38
- [x] https://github.com/mojaloop/central-event-processor/pull/34
- [x] https://github.com/mojaloop/email-notifier/pull/27
- [x] https://github.com/mojaloop/central-ledger/pull/302
~- [ ] https://github.com/mojaloop/helm/pull/197~ Revised in subsequent PR
- [x] https://github.com/mojaloop/helm/pull/199
- [x] https://github.com/mojaloop/ml-api-adapter/pull/124
- [ ] https://github.com/mojaloop/central-settlement/pull/79

## **Follow-up**:
- N/A

**Dependencies**:
- [ ] https://github.com/mojaloop/project/issues/817 (currently blocking changes to `central-settlement`)

## **Accountability**:
- Owner: TBC
- QA/Review: TBC

**Comment by lewisdaly** [comment]
A few questions about how the health checks:

1. How are they currently being used? I gather that Kubernetes is consuming health checks for each of the services to check for liveness. Are there any other services that are consuming the current health checks?
2. Is backwards compatibility a concern? (related to Question 1) Perhaps it would make sense as part of the design to have a `detailed` url Parameter - eg. `/health?detailed=true` would return the new model, but `/health` would remain unchanged.
3. Are there any security concerns I should be worried about? Typically health check endpoints are public, so I guess we should avoid doing any heavy work on this endpoint (in case of DDOS) or exposing too much information.

**Comment by mdebarros** [comment]
See my inline responses:

> A few questions about the health checks:
> 
> 1. How are they currently being used? I gather that Kubernetes is consuming health checks for each of the services to check for liveness. Are there any other services that are consuming the current health checks?

Currently only being used for `readiness` & `liveness` probes on Kubernetes.

> 2. Is backwards compatibility a concern? (related to Question 1) Perhaps it would make sense as part of the design to have a `detailed` url Parameter - eg. `/health?detailed=true` would return the new model, but `/health` would remain unchanged.

There should be no backward compatibility issues as to my knowledge as only the HTTP-200 status code response is used by Kubernetes, with no interrogation to the body itself. Also, keep in mind that you are also adding new fields to the body itself and not removing/changing the existing fields....this would imply that it should be backward compatible regardless. Thus the `?detailed=true` mode is probably unnecessary.

> 3. Are there any security concerns I should be worried about? Typically health check endpoints are public, so I guess we should avoid doing any heavy work on this endpoint (in case of DDOS) or exposing too much information.

That is a fair comment.
- Security: The idea is that all Mojaloop services should be accessible via a secured API Gateway where operations such as `/health` should not necessarily be exposed publically. 
- DDOS/Performance: 
     - I would think not including the `latency` information would be prudent here as this will most likely be a target
     - The `connectivity checks` may also prove to be an issue, but most tightly coupled back-ends (e.g. db, kafka, etc) are connected via a client library that intermittently provides connectivity checks already. I would thus expect this to not impact performance.

With regard to question 2 & 3 above, that got me thinking that we could go with an alternative approach as follows:

1. `/health` mode operates as it currently with the original `status` field, but with added `versionNumber`, `uptime`, `started` as per your suggested design. It does NOT interrogate the back-ends services for a status. <-- this could be used for a quick/often service health checks such as the k8s `liveliness` probes
1. `/health?detailed=true` mode operates fully as per your design including the back-end checks and would include the `serviceHealth` list in addition to the response. <-- this may be better suites for k8s `readiness` probes.

Let me know if you have any additional questions, and also what you think about my alternative thoughts to your design.

**Comment by lewisdaly** [comment]
Yeah I like your suggestions for `/health` and `/health?detailed=true` above.

My only thought would be on how we calculate whether or not a service is `DOWN` when hitting `/health` _i.e. not the detailed health check_. Should we still mark the service as down if we don't ask for a detailed health check, but sub-services are down?

Also, we can leave out the latency info for now.

**Comment by lewisdaly** [comment]
@elnyry said not to worry about the sidecar for now, since it will be deprecated soon

**Comment by elnyry** [comment]
sorry, I meant refactored :-)

**Comment by mdebarros** [comment]
> Yeah I like your suggestions for `/health` and `/health?detailed=true` above.
> 
> My only thought would be on how we calculate whether or not a service is `DOWN` when hitting `/health` _i.e. not the detailed health check_. Should we still mark the service as down if we don't ask for a detailed health check, but sub-services are down?

If we can do it without any performance impacts/issues, then I would say we should include the sub-services as part of it. Otherwise, we may have to introduce a new status to indicate that the API is up but we have not checked the sub-services.

> Also, we can leave out the latency info for now.

That is probably not a bad idea. For interest sake, how did you think of implementing the latency checks?

**Comment by lewisdaly** [comment]
I agree that keeping it consistent is probably best and easiest.

Based on some implementation I was working on today with the `central-ledger`, there wasn't any real overhead with monitoring the sub-services, as I used the existing libraries that had already established connections.

Regarding the latency, I found some example health checks that had used this and was inspired by that. Basically I think we would need to implement some sort of ping with a timer for each service, but this would then create overhead that we are trying to avoid, as mentioned above, so perhaps we are better off avoiding it.

Another thought; for monitoring latency when connecting to Kafka for example, we could also get in the situation where we have multiple topics on different hosts (at least according to how the defaults.json is laid out), which would make things a lot harder to manage and make overhead even greater for each check, so I think using the existing libraries is the best option for now.
